jobs: set clusterID and clusterVersion during schedule creation

msbutler opened the pull request:
upgrade: pass ClusterID to tenantDeps
Release note: None

----

jobs: set clusterID and clusterVersion during schedule creation
This patch adds two new fields to the ScheduleDetails proto, apart of the
schedule record:
- ClusterID: the cluster that created the schedule. Note that in a future PR,
  we will allow the backup schedule to alter this field on resume.
- CreationClusterVersion: the cluster version the schedule was created on.

This patch ensures that these fields must be set during schedule creation.

In the future, schedules after RESTORE and C2C cutover will pause themselves to
prevent conflicting with schedules potentially running on the backup/source
cluster.

Informs https://github.com/cockroachdb/cockroach/issues/108028

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/111374)
<!-- Reviewable:end -->

adityamaru commented on pkg/upgrade/upgrades/ensure_sql_schema_telemetry_schedule.go:
```diff
@@ -17,15 +17,20 @@ import (
 	ctx context.Context, cs clusterversion.ClusterVersion, d upgrade.TenantDeps,
 ) error {
 	return d.DB.Txn(ctx, func(ctx context.Context, txn isql.Txn) error {
+		// Since this Schedule Creation job is expected to fail before the schedule
```
i'm not sure I follow this, isn't this expected to fail only if a telemetry schedule exists? In other cases this is run as a migration to auto heal the cluster right?

msbutler replied:
good point. I rephrased the comment.

adityamaru commented on pkg/jobs/scheduled_job.go:
```diff
@@ -462,7 +464,18 @@ func (s scheduledJobStorageTxn) Create(ctx context.Context, j *ScheduledJob) err
+	if j.rec.ScheduleDetails.CreationClusterVersion.Equal(clusterversion.ClusterVersion{}) {
+		return errors.New("scheduled job created without a cluster version")
+	}
+	if j.rec.ScheduleDetails.CreationClusterVersion.Less(clusterversion.ByKey(clusterversion.V23_2)) {
```
is this strictly necessary? Won't older binaries just not have those fields on the proto they unmarshal the column into. What would make me feel confident here is a mixed version logic test that:

- creates a schedule on master and loads it on 23.1
- creates a schedule on 23.1 and loads it on master

msbutler replied:
good callout. I added a mixed version test for your perusal. 

I version gated the _writing_ of the clusterID and version because in a mixed version state, older nodes _could_ overwrite these new fields. e.g. `ALTER BACKUP SCHEDULE ...  SET SCHEDULE OPTION on_previous_running = 'wait';`  from an old node in a mixed version state would overwrite these new fields.  It just seems cleaner to me for these fields to only populate once we're fully upgraded.

adityamaru commented on pkg/jobs/scheduled_job_test.go:
```diff
@@ -121,3 +125,50 @@ func TestPauseUnpauseJob(t *testing.T) {
+	// Init a mixed version schedule
+	sjMixed := h.newScheduledJob(t, "mixed_test_job", "test sql")
+
+	// In a mixed version state, clusterID and clusterVersion fields do not populate.
```
Hm so this is not actually exercising the `protoutil.Unmarshal` call that a pre-23.2 binary would make on a scheduled job created on 23.2+, which is why I don't think this is enough. This is simply testing that we zero out the fields which as my other comment indicates I'm not sure is needed. When an older version binary attempts to unmarshal protobuf bytes written by a newer version, the added fields will simply be ignored.

Can we maybe replace this with a mixed-version logictest instead?

msbutler replied:
mixed version test added. Given my above comment on version gating the writing of these fields, I still think this test case is necessary.

adityamaru reviewed (changes requested):
Overall LGTM, just a few comments

msbutler commented on pkg/jobs/jobspb/schedule.proto:
```diff
@@ -44,6 +47,19 @@ message ScheduleDetails {
   // How to handle failed jobs.
   ErrorHandlingBehavior on_error = 2;
+
+  // ClusterID is populated at schedule creation with the ClusterID, in case a
```
I'm going to keep what i have for now, as I don't like that it would be up to the author to maintain the invariant. Depending on how I implement the follow up PR, I may change the comment to your suggestion.

msbutler committed:
upgrade: pass ClusterID to tenantDeps

Release note: None

msbutler committed:
jobs: set clusterID and clusterVersion during schedule creation

This patch adds two new fields to the ScheduleDetails proto, apart of the
schedule record:

ClusterID: the cluster that created the schedule. Note that in a future PR,
we will allow the backup schedule to alter this field on resume.
CreationClusterVersion: the cluster version the schedule was created on.
This patch ensures that these fields must be set during schedule creation.

In the future, schedules after RESTORE and C2C cutover will pause themselves to
prevent conflicting with schedules potentially running on the backup/source
cluster.

Informs #108028

Release note: None

fqazi reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> , just one minor comment

> Reviewed 27 of 27 files at r2, 26 of 26 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/111374#-:-NfvftGB3B0yJBe1ju-0:b13jg4p)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @adityamaru, @dt, @miretskiy, and @msbutler)

___
*[`pkg/jobs/jobstest/utils.go` line 151 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/111374#-NfvfWJ8EX8Jy2D74lqV:-NfvfWJ90cLz8wWkz-it:b-oac0re) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b1fe64bd740a35cfc4118fabf1d61ebd56aadda/pkg/jobs/jobstest/utils.go#L151)):*
> ```Go
> 
> // DummyClusterVersion is used while instantiating dummy schedules
> var DummyClusterVersion = clusterversion.ClusterVersion{Version: clusterversion.ByKey(clusterversion.V23_2)}
> ```

Maybe just use TestingBinaryVersion here?


<!-- Sent from Reviewable.io -->

msbutler commented:
TFTRs!!

bors r=adityamaru
